lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -170,6 +171,113 @@ static void revertDependentTypeLoc(TypeLoc &tl) {
+  TypeLoc constraintTypeLoc(repr->getConstraint());
+  // Pass along the error type if resolving the repr failed.
+  bool validationError
+    = tc.validateType(constraintTypeLoc, resolution, options);
+  auto constraintType = constraintTypeLoc.getType();
+  if (validationError)
+    return constraintType;
+  
+  // Error out if the constraint type isn't a class or existential type.
+  if (!constraintType->getClassOrBoundGenericClass()
+      && !constraintType->isExistentialType()) {
+    tc.diagnose(repr->getConstraint()->getLoc(),
+                diag::opaque_type_invalid_constraint);
+    return constraintTypeLoc.getType();
+  }
+
+  // Create a generic signature for the opaque environment. This is the outer
+  // generic signature with an added generic parameter representing the opaque
+  // type and its interface constraints.
+  SmallVector<GenericTypeParamType *, 4> interfaceGenericParams;
+  SmallVector<Requirement, 4> interfaceRequirements;
+
+  auto originatingDC = originatingDecl->getInnermostDeclContext();
+  auto outerGenericSignature = originatingDC->getGenericSignatureOfContext();
+  if (outerGenericSignature) {
+    std::copy(outerGenericSignature->getGenericParams().begin(),
+              outerGenericSignature->getGenericParams().end(),
+              std::back_inserter(interfaceGenericParams));
+    std::copy(outerGenericSignature->getRequirements().begin(),
+              outerGenericSignature->getRequirements().end(),
+              std::back_inserter(interfaceRequirements));
+  }
+
+  unsigned returnTypeDepth = 0;
+  if (!interfaceGenericParams.empty())
+    returnTypeDepth = interfaceGenericParams.back()->getDepth() + 1;
+  auto returnTypeParam = GenericTypeParamType::get(returnTypeDepth, 0,
+                                                   tc.Context);
+  interfaceGenericParams.push_back(returnTypeParam);
+
+  if (constraintType->getClassOrBoundGenericClass()) {
+    // Use the type as a superclass constraint.
+    interfaceRequirements.push_back(Requirement(RequirementKind::Superclass,
+                                              returnTypeParam, constraintType));
+  } else {
+    auto constraints = constraintType->getExistentialLayout();
+    if (auto superclass = constraints.getSuperclass()) {
+      interfaceRequirements.push_back(Requirement(RequirementKind::Superclass,
+                                                  returnTypeParam, superclass));
+    }
+    for (auto protocol : constraints.getProtocols()) {
+      interfaceRequirements.push_back(Requirement(RequirementKind::Conformance,
+                                                  returnTypeParam, protocol));
+    }
+    if (auto layout = constraints.getLayoutConstraint()) {
+      interfaceRequirements.push_back(Requirement(RequirementKind::Layout,
+                                                  returnTypeParam, layout));
+    }
+  }
+  
+  auto interfaceSignature = GenericSignature::get(interfaceGenericParams,
```

Comment:
It might be cleaner if you use a GSB instead; then you can add a requirement with a class or any kind of existential in it and it will flatten it into the right thing. In particular, there's no guarantee that the order in which you're adding stuff above is the right order to get a canonical generic signature